accepted/install-locations.md

```diff
@@ -0,0 +1,216 @@
+
+**Important:** There's a difference in search algorithm for `hostfxr` and for frameworks/SDKs. For `hostfxr` only the first available location is considered (and the highest version from that location is selected). For frameworks/SDKs **all** locations are considered and the best match for the requested version is selected from all those locations.
+
+Search locations:
+* the location of the `hostfxr` used. See the `hostfxr` search algorithm above. This means that `DOTNET_ROOT` has effect on framework and SDK search as well since the `hostfxr` search will consider it.  
+*If the `hostfxr` is found in the `DOTNET_ROOT` location (should happen pretty much always when the variable is non-empty), then that location will be part of the search locations for frameworks/SDKs.*
+* If multi-level lookup is enabled (by default it is, can be disabled via `DOTNET_MULTILEVEL_LOOKUP=0` environment variable), then the following locations are searched as well:
+  * :new: global registered install location  
+    **We've committed to support this on Windows for .NET Core 3.0**
+    * Windows - `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` (32-bit view)  
+    * Linux/macOS - `/etc/dotnet/install_location`
+  * default global location, in the `/host/fxr/<version>` subdirectory
+    * Windows (32-bit process on 64-bit Windows) - `%ProgramFiles(x86)%\dotnet`
+    * Windows (everything else) - `%ProgramFiles%\dotnet`
+    * :new: Linux - `/usr/share/dotnet`
+    * :new: macOS - `/usr/local/share/dotnet`
+* Else multi-level lookup is disabled, no other search locations are considered.
+
+The reason for including the default global location is so that 3.0 host can find 2.* installs since those are not registered in the system. All 3.0 and higher installs should be registered.
+
+*Note: The above means that on Linux/macOS in 2.2 the multi-level lookup was effectively non-functional, as there were no search paths for it. Part of the 3.0 changes is to change that and make Linux/macOS work similarly to Windows in this case as well.*  
+*Note: The multi-level lookup feature in its current form seems to be rather confusing and there are discussions around phasing it out. See dotnet/core-setup#3606. If there's no change in that area in .NET Core 3.0 we should do what's proposed here as it makes all platforms consistent.*
+
+
+## Open questions
+
+### Make muxer less special
+As mentioned above, the muxer has certain specific behavior:
+* It only looks next to itself when searching for `hostfxr`.
+* It ignores `DOTNET_ROOT` environment variable.
+
+#### Framework and SDK search
+I think the muxer should use `DOTNET_ROOT` to resolve frameworks and SDKs just like all the other hosts (the actual implementation mechanism aside). The current discrepancy is simply unexpected and makes certain scenarios unnecessarily complex (for example VS can't really use private installs because of this, since it always uses the muxer from path).
```

Comment:
Note that VS worked fine with private installs when they were put on PATH. Until apphost became defaullt for F5 in 3.0 and now you need DOTNET_ROOT as well.

This has been a topic of much debate in the past. People have argued various things:

1. Should muxer respect DOTNET_ROOT as you propose, but also:

2. Should muxer export DOTNET_ROOT to children. One thing that people find surprising is `/some/private/dotnet X` launching `dotnet Y` and the latter can't find framework or SDK in /some/private.

3. Should CLI do (2) if muxer won't.

The problem with DOTNET_ROOT is that it is a hammer. Setting it can fix some things and break others. The fact that you initiated a process chain with some dotnet probably means that the frameworks and sdks in that location should be considered. We know from experience that folks consider this surprising when it isn't. On the other hand, there can be other things in the chain that need the global locations. For example, in a build, I run a global tool built against 2.1 and I have 2.1 globally, but I also run a freshly built 3.0 apphost in the same process chain and I only have 3.0 in a private location. 

For this reason, we have been reluctant to guess and set DOTNET_ROOT. Except that /some/private/dotnet run was completely broken by apphost, so now we set it in `dotnet run` but that has been thought of as a hack pending the outcome here.